### ISSUE TITLE
dbus-mqtt: reduce Venus cpu load

### ISSUE THREAD

**Issue by mpvader** [victronenergy/venus#189]
dbus-mqtt uses a lot of cpu. And, because its a separate (python) process that subscribes to the dbus, it negatively impacts how many devices can be connected to Venus; since then, besides systemcalc, vrmlogger, and-so-forth, also dbus-mqtt gets some extra load.

Various features are underway which will change dbus-mqtt from a thing used by some people into a commonly enabled service. The html5-app is one example, probably VictronConnect over MQTT will use it too; and then there is the pending live-update feature for VRM.

Conclusion: its time to improve this.

Some ideas:
- leave existing functionality in dbus-mqtt as it is, and add sending batched system updates to it. (update 28-jan-2019: I'm not sure what I meant when typing that.. )
- rather than run as its own process, add the functionality to another one that already runs and listens to D-Bus, such as systemcalc or vrmlogger.

A simple test that needs doing is disabling the mqtt sending part and running it; to see how much load comes from just listening to the D-Bus.

Repo: https://github.com/victronenergy/dbus-mqtt

**Comment by izak** [comment]
After some time benchmarking and trying different things, it appears the load comes down to roughly this:

* Around 5% is used to react to dbus events and put the result into a dictionary.
* Around 5%-10% of the CPU is used to json encode a little `{value: xxx}` dictionary.
* Maybe 5% or so is spent unwrapping dbus types into python types. 
* The rest of the time is used to send data to the mqtt daemon, that is more than half, as much as 80%.

So while small improvements are possible in the first three cases, the bulk of the poor performance is down simply to the volume of data that has to be repeated from dbus onto mqtt.

One thing that really helps is to put in a queue mechanism between the reception of a value from dbus (either from PropertiesChanged or request/scanning) and the actual sending of the response, and to add a bit of rate limiting to it: if you have multiple PropertiesChanged events for the same topic, only send the most recent one.

The important question at this point is how real-time it needs to be. Would it matter if there is occasionally a small 2 second (or 1.5 seconds, or 1 second) delay between a dbus event and the resultant mqtt message?

**Comment by wiebeytec** [comment]
> The rest of the time is used to send data to the mqtt daemon, that is more than half, as much as 80%.

Can you clarify this a little? What does 'sending to the mqtt deamon' mean? I find it suspect that merely sending bytes to a TCP socket takes that long. Plus, is it 'time' (waiting for `select()`), or CPU cycles?

Is there also a lot of Mosquitto CPU use? Does Venus use Mosquitto 1.5, built with E_POLL? (alhtough, because there's only one client to the daemon it may not make a difference, but you never know).

**Comment by izak** [comment]
> Can you clarify this a little?

The numbers are a very crude estimate. I looked pretty much at the amount of CPU time reported by top, and made an estimate as to where time is spent (and hence where I should spend time trying to optimise).

My crude estimate is 1) a small amount of time in reacting to dbus PropertiesChanged and storing them, 2) another small amount in json encoding and wrapping/unwrapping dbus types, and 3) the bulk of it just shovelling data into mosquitto.

Most of the time is spent, amazingly, in socket.send, which I assume must be because of the sheer amount of data passing through here. After adding a rate limiter, the CPU load (again, as reported by top) went down quite a bit.

mosquitto shows quite a bit of load when dbus-mqtt sends all 1000+ topics into it (the first time you connect to mqtt and send a keepalive), but does not appear to be a major consumer the rest of the time.

I did try a few optimisations on json and dbus encoding, but eventually found the gains too small to be worth it.
